core/services/blockheaderfeeder/delegate.go

```diff
@@ -0,0 +1,219 @@
+package blockheaderfeeder
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/chains/evm"
+	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/batch_blockhash_store"
+	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/blockhash_store"
+	v1 "github.com/smartcontractkit/chainlink/core/gethwrappers/generated/solidity_vrf_coordinator_interface"
+	v2 "github.com/smartcontractkit/chainlink/core/gethwrappers/generated/vrf_coordinator_v2"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/blockhashstore"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/multierr"
+)
+
+var _ job.ServiceCtx = &service{}
+
+type Delegate struct {
+	logger logger.Logger
+	chains evm.ChainSet
+	ks     keystore.Eth
+}
+
+func NewDelegate(
+	logger logger.Logger,
+	chains evm.ChainSet,
+	ks keystore.Eth,
+) *Delegate {
+	return &Delegate{
+		logger: logger,
+		chains: chains,
+		ks:     ks,
+	}
+}
+
+// JobType satisfies the job.Delegate interface.
+func (d *Delegate) JobType() job.Type {
+	return job.BlockHeaderFeeder
+}
+
+// ServicesForSpec satisfies the job.Delegate interface.
+func (d *Delegate) ServicesForSpec(jb job.Job) ([]job.ServiceCtx, error) {
+	if jb.BlockHeaderFeederSpec == nil {
+		return nil, errors.Errorf("Delegate expects a BlockHeaderFeederSpec to be present, got %+v", jb)
+	}
+
+	chain, err := d.chains.Get(jb.BlockHeaderFeederSpec.EVMChainID.ToInt())
+	if err != nil {
+		return nil, fmt.Errorf(
+			"getting chain ID %d: %w", jb.BlockHeaderFeederSpec.EVMChainID.ToInt(), err)
+	}
+
+	if jb.BlockHeaderFeederSpec.LookbackBlocks < int32(chain.Config().EvmFinalityDepth()) {
+		return nil, fmt.Errorf(
+			"lookbackBlocks must be greater than or equal to chain's finality depth (%d), currently %d",
+			chain.Config().EvmFinalityDepth(), jb.BlockHeaderFeederSpec.LookbackBlocks)
+	}
+
+	keys, err := d.ks.EnabledKeysForChain(chain.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "getting sending keys")
+	}
+	if len(keys) == 0 {
+		return nil, fmt.Errorf("missing sending keys for chain ID: %v", chain.ID())
+	}
+	if err := CheckFromAddressesExist(jb, d.ks); err != nil {
+		return nil, err
+	}
+	fromAddresses := jb.BlockHeaderFeederSpec.FromAddresses
```

Comment:
makes sense. I ended up moving the `roundRobin` call to outside of all the storage calls. This allows us to specify multiple `fromAddresses` in case a sending key stops working. At the same time, this should guarantee that only a single key is used for each run of the feeder so that batches of store operations are serialized.